### PR TITLE
Improve "associated type not found" diagnostics

### PR DESCRIPTION
**Pull request by ShE3py**
```rs
use core::ops::Deref;

fn foo<T>() where T: Deref<Output = u32> {}
```

Before:
```
error[E0220]: associated type `Output` not found for `Deref`
 --> E0220.rs:5:28
  |
5 | fn foo<T>() where T: Deref<Output = u32> {}
  |                            ^^^^^^ associated type `Output` not found
```

After:
```
error[E0220]: associated type `Output` not found for `Deref`
 --> E0220.rs:5:28
  |
5 | fn foo<T>() where T: Deref<Output = u32> {}
  |                            ^^^^^^ help: `Deref` has the following associated type: `Target`
```

---

@rustbot label +A-diagnostics +D-papercut

### CODE REVIEW

File: tests/ui/associated-type-bounds/missing-trait-bound-for-assoc-fails.stderr [hunk 1]
```diff
@@ -8,7 +8,9 @@ error[E0220]: associated type `Item` not found for `M`
   --> $DIR/missing-trait-bound-for-assoc-fails.rs:4:8
    |
 LL |     M::Item: Temp,
-   |        ^^^^ associated type `Item` not found
+   |        ^^^^
+   |
+   = help: `M` has no associated type, try removing `Item`
```
Comment (compiler-errors) [inline tests/ui/associated-type-bounds/missing-trait-bound-for-assoc-fails.stderr:13]: This suggestion doesn't make sense imo.

File: tests/ui/associated-types/associated-types-path-1.stderr [hunk 2]
```diff
@@ -2,7 +2,9 @@ error[E0220]: associated type `A` not found for `T`
   --> $DIR/associated-types-path-1.rs:10:26
    |
 LL | pub fn f1<T>(a: T, x: T::A) {}
-   |                          ^ associated type `A` not found
+   |                          ^
+   |
+   = help: `T` has no associated type, try removing `A`
```
Comment (compiler-errors) [inline tests/ui/associated-types/associated-types-path-1.stderr:7]: Nor does this suggestion make sense.

Comment (compiler-errors) [reply tests/ui/associated-types/associated-types-path-1.stderr:7]: We already suggest what traits provide `A` below. We should just remove this help.

File: tests/ui/associated-consts/assoc-const-eq-missing.stderr [hunk 3]
```diff
@@ -2,19 +2,25 @@ error[E0220]: associated type `Z` not found for `Foo`
   --> $DIR/assoc-const-eq-missing.rs:15:16
    |
 LL | fn foo1<F: Foo<Z=3>>() {}
-   |                ^ associated type `Z` not found
+   |                ^
+   |
+   = help: `Foo` has no associated type, try removing `Z`
```
Comment (compiler-errors) [inline tests/ui/associated-consts/assoc-const-eq-missing.stderr:7]: I don't think the right suggestion for a user proving an associated type bound that can't be found is to remove it.

**Review by compiler-errors — CHANGES_REQUESTED** [review]

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 4]
```diff
@@ -197,7 +201,34 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+        // If we still couldn't find any associated type, just list them all.
+
+        if all_candidate_names.is_empty() {
+            err.help(format!(
+                "`{ty_param_name}` has no associated type, try removing `{assoc_name}`"
+            ));
+            return err.emit();
+        }
+
+        let msg = if all_candidate_names.len() > 1 {
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:213]: I don't think we should suggest anything if there's >1 candidate.

File: compiler/rustc_hir_analysis/src/astconv/errors.rs [hunk 5]
```diff
@@ -197,7 +201,34 @@ impl<'o, 'tcx> dyn AstConv<'tcx> + 'o {
...
+            format!("`{ty_param_name}` has the following associated types")
+        } else {
+            format!("`{ty_param_name}` has the following associated type")
+        };
+
+        let applicability = if self.tcx().features().associated_type_defaults {
+            Applicability::Unspecified // `type A = Self::B` would suggest `type A = Self::A`
```
Comment (compiler-errors) [inline compiler/rustc_hir_analysis/src/astconv/errors.rs:220]: What case is this protecting against?

Comment (ShE3py) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:220]: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2021&gist=848b04b36dea54f6539f37afd21fcd89
The compiler errors on `type A = Self::A`.

File: tests/ui/associated-types/associated-types-path-1.stderr (hunk 2 above)
Comment (ShE3py) [reply tests/ui/associated-types/associated-types-path-1.stderr:7]: Didn't thought much about it but yeah, if the user tries to use an associated type the compiler shouldn't tell them to remove it.

File: compiler/rustc_hir_analysis/src/astconv/errors.rs (hunk 4 above)
Comment (ShE3py) [reply compiler/rustc_hir_analysis/src/astconv/errors.rs:213]:
```rs
trait Qux {
    type A;
    type B;
}

fn quux<T>() where T: Qux<A = u32, ZZ = u32> {}
```
gives
```
error[E0220]: associated type `ZZ` not found for `Qux`
  --> E0220.rs:14:36
   |
14 | fn quux<T>() where T: Qux<A = u32, ZZ = u32> {}
   |                                    ^^
   |
help: `Qux` has the following associated types
   |
14 | fn quux<T>() where T: Qux<A = u32, A = u32> {}
   |                                    ~
14 | fn quux<T>() where T: Qux<A = u32, B = u32> {}
   |                                    ~
```
I think this is still helpful?

**Review by compiler-errors — APPROVED** [review]

**Pull request merged**